### ISSUE TITLE
Index saved as file per data file

### ISSUE THREAD

**Issue by rjasonadams** [prologic/bitcask#108]
Hello,

Have you considered saving the index to a file per data file? Since the data files are immutable once finished the index would only have to be written once per data file. I think this would be similar to how hint files work in Riak's implementation? The generation of the saved index could be done in the background (by reading over the data file?) or it could be triggered when the current data file is closed and a new one started.

**Comment by prologic** [comment]
Yeah this implemtnation is a little different now; We used to do that... But now we're using an Adaptive Radix tree as the "index" so to speak and its a single data structure in memory that basically just does the job of the index as per the spec holding keys to items (locations of the value on disk).

**Comment by rjasonadams** [comment]
What lead me to consider this is the way the index is recreated if missing. It looks to read over the data file and generates the index. Since our use case has a large data set (16TB) I cringed at how long this would take if we lost the index. It seems like creating an index per file would be very similar to how a missing index is created. If the index for the file exists, loading it (instead of reading the entire data file) would be quick.

**Comment by prologic** [comment]
Yeah I understand your worry. But... isn't the end result the same? Or are you saying maybe its better in general to split out the "on-disk index" files so if only one is lost we only have to re-create that one and read only that one datafile?

**Comment by prologic** [comment]
So I _might_ have to re-consider some things here; because you bring up an interesting point. We no longer write our `.hint` files; just one index file which really is the in-memory Radix Tree.

If we also solve your other issue #109 -- We should probably also go back to writing out `.hint` files per data file as part of that operation and use those to re-create the "index" if its ever lost, corrupted. This would save you from having to re-read the entire on-disk dataset for a missing index.

Happy for a PR to solve both #108 and #109
